### PR TITLE
BIOIN-2482 Refactor CNV filtering models, filter CNV duplications

### PR DESCRIPTION
**Pull request by ilyasoifer**
This PR finishes refactoring of the CNV pipeline. 

1. Training pipeline now has a capability to train cnv filtering model
2. Removed temporary solution of train_cnv_filtering_model
3. Added specific script that prepares ground truth for CNV filtering

Currently we converged to the following structure
1. Not to collapse the input CNVs
2. Merge small CN.MOPS duplications
3. Annotate VCF with UG-CNV-LCR regions, but they are not used in the training model
4. The model ignores CNV type
5. Merge variants that pass filter post-filtering

---

> [!NOTE]
> Refactors CNV utilities and pipelines, introduces CNV training-prep and CNV-aware filtering/modeling, enhances VCF merge/annotation and SV comparison workflows, and updates tests/resources accordingly.
> 
> - **CNV Module**:
>   - Centralize VCF tag registries in `ugbio_cnv/cnv_vcf_consts.py`; switch gap metric to `GAP_PERCENTAGE` and add `REGION_ANNOTATIONS`.
>   - Refactor `cnmops_utils.py`: expose helpers to annotate/aggregate/write CNVs (`annotate_bed`, `aggregate_coverages`, `aggregate_annotations_in_df`, `add_ids`, `write_cnv_vcf`); remove legacy CLI/data-prep code.
>   - Enhance `combine_cnv_vcf_utils.merge_cnvs_in_vcf` with options (`ignore_sv_type`, `ignore_filter`, `pick_best`, `do_not_merge_collapsed_filtered`), aggregate extra INFO fields, drop collapse helper INFO, and ensure sorting; `write_vcf_records_with_source` now sets `PASS`.
>   - Update `combine_cnmops_cnvpytor_cnv_calls.py`: add `merge_records` subcommand, rename gap annotation to `GAP_PERCENTAGE`, adjust CNMOPS short-dup filtering, remove unused jalign processing.
>   - Update `process_cnmops_cnvs.py` and `filter_sample_cnvs.py` to use new utilities.
>   - Docker: set `PYTOR_DATA_DIR` and precreate dir before `cnvpytor -download`.
> - **Filtering/Training**:
>   - Add CNV training prep CLI `training_prep_cnv_pipeline.py` and `training_prep.training_prep_cnv`; remove old standalone `train_cnv_filtering_model` path and tests.
>   - Extend transformers and modeling to `VcfType.CNV` (new feature encoders, CNV-specific XGBoost params, grouping by `SVTYPE`).
>   - `filter_variants_pipeline.py`: add `--overwrite_qual_tag` and configurable `--decision_threshold`, stronger validation.
> - **SV Comparison**:
>   - `SVComparison`: support `ignore_type` and `skip_collapse`; when `ignore_filter`, strip filters via `VcfUtils.remove_filters`; more robust `svlen` handling.
> - **Core Utils**:
>   - `VcfUtils`: add `remove_filters`, support `pick_best` in `collapse_vcf`, and specify temp dir in `sort_vcf`/`concat_vcf`.
> - **Tests/Resources**:
>   - Add/expand unit and system tests for CNV merging, gap/region annotations, CNV training prep, model training, and comparison pipeline; update LFS test resources; remove legacy tests tied to deleted code.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 394da49731642905c804728ea8f46cdaef46c522. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by ilyasoifer** [comment]
/wdl-tests workflows=SingleSampleCnmopsCNVCalling,SingleSampleCNVpytorCalling

File: src/cnv/Dockerfile [hunk 1]
```diff
@@ -99,6 +99,8 @@ COPY --from=build ${OUT_DIR} ${OUT_DIR}
 RUN pip install ${OUT_DIR}/*.whl && rm -rf ${OUT_DIR}
 
 #download cnvpytor additional necessary files (described here: https://github.com/abyzovlab/CNVpytor?tab=readme-ov-file#install-using-pip)
+ENV PYTOR_DATA_DIR=/usr/share/cnvpytor_data
+RUN mkdir -p $PYTOR_DATA_DIR
```
Comment (TammushB) [inline src/cnv/Dockerfile]: is the data downloads to the PYTOR_DATA_DIR ? is that automatic ?

File: src/cnv/ugbio_cnv/cnv_consts.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+# (after converting table to BED the names are as in the VCF)
+INFO_TAG_REGISTRY: dict[str, tuple[str, int | str, str, str, str]] = {
+    "CNV_calls_source": (
+        "CNV_SOURCE",
+        ".",
+        "String",
+        "the tool that called this CNV. can be combination of: cn.mops, cnvpytor, gridss",
+        "INFO",
+    ),
+    "CNV_SOURCE": (
```
Comment (TammushB) [inline src/cnv/ugbio_cnv/cnv_consts.py]: isn't this duplication for CNV_calls_source ?

File: src/cnv/ugbio_cnv/cnv_consts.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        "INFO",
+    ),
+    "CNMOPS_COHORT_STDEV": (
+        "CNMOPS_COHORT_STDEV",
+        1,
+        "Float",
+        "Standard deviation of coverage in the CNV region across the cohort (cn.mops)",
+        "INFO",
+    ),
+    "GAP_PERC": ("GAP_PERC", 1, "Float", "Fraction of N bases in the CNV region from reference genome", "INFO"),
```
Comment (TammushB) [inline src/cnv/ugbio_cnv/cnv_consts.py]:
```suggestion
    "GAP_PERCENTAGE": ("GAP_PERCENTAGE", 1, "Float", "Fraction of N bases in the CNV region from reference genome", "INFO"),
```

Comment (TammushB) [reply src/cnv/ugbio_cnv/cnv_consts.py]: explicit is always better ;)

File: src/cnv/ugbio_cnv/combine_cnmops_cnvpytor_cnv_calls.py [hunk 4]
```diff
@@ -94,6 +93,19 @@ def __parse_args_annotate_regions(parser: argparse.ArgumentParser) -> None:
     )
 
 
+def __parse_args_merge_records(parser: argparse.ArgumentParser) -> None:
+    """Add arguments specific to the merge records tool."""
+    parser.add_argument("--input_vcf", help="Input VCF file with CNV calls to merge", required=True, type=str)
+    parser.add_argument("--output_vcf", help="Output VCF file with merged CNV calls", required=True, type=str)
+    parser.add_argument(
+        "--distance",
+        help="Distance threshold for merging CNV segments (default: 0)",
```
Comment (TammushB) [inline src/cnv/ugbio_cnv/combine_cnmops_cnvpytor_cnv_calls.py]: I think in terra we used default of 1500.

File: src/cnv/ugbio_cnv/combine_cnv_vcf_utils.py [hunk 5]
```diff
@@ -264,43 +283,112 @@ def merge_cnvs_in_vcf(input_vcf: str, output_vcf: str, distance: int = 1000) ->
...
                     record.info["SVLEN"] = (record.stop - record.start,)
+                collapse_info_fields = ["CollapseId", "NumCollapsed", "NumConsolidated"]
+                for c in collapse_info_fields:
+                    if c in record.info:
+                        del record.info[c]
                 vcf_out.write(record)
     mu.cleanup_temp_files(temporary_files)
+
+
+AGGREGATION_ACTIONS = {
```
Comment (TammushB) [inline src/cnv/ugbio_cnv/combine_cnv_vcf_utils.py]: maybe this block should be in the beginning of the script or private inside merge_cnvs_in_vcf function.

File: src/cnv/ugbio_cnv/cnmops_utils.py [hunk 6]
```diff
@@ -1,118 +1,16 @@
...
-        None,
-        None,
-        "Duplication length is shorter than the defined threshold in cn.mops calls.",
-        "FILTER",
-    ),
-}
+warnings.filterwarnings("ignore")
 
 
 def add_vcf_header(sample_name: str, fasta_index_file: str) -> pysam.VariantHeader:
```
Comment (TammushB) [inline src/cnv/ugbio_cnv/cnmops_utils.py]: is this the right place for this function? it is not cnmops specific .

File: src/cnv/ugbio_cnv/cnmops_utils.py [hunk 7]
```diff
@@ -525,75 +348,149 @@ def write_cnv_vcf(outfile: str, cnv_df: pd.DataFrame, sample_name: str, fasta_in
...
-    parser.add_argument("--verbosity", help="Verbosity: ERROR, WARNING, INFO, DEBUG", required=False, default="INFO")
+    # Process each coverage annotation file
+    for sample_name, operation, bed_file_path in coverage_annotations:
+        # Read the coverage annotation bed file
+        cov_df = pd.read_csv(bed_file_path, sep="\t", header=None)
 
-    args = parser.parse_args(argv[1:])
-    logger.setLevel(getattr(logging, args.verbosity))
+        # Assign column names for sorting
+        cov_df.columns = ["chrom", "start", "end", "name", "cov"]
```
Comment (TammushB) [inline src/cnv/ugbio_cnv/cnmops_utils.py]: by "name" you mean "sample_name" ?

File: src/filtering/ugbio_filtering/transformers.py
Comment (TammushB) [inline src/filtering/ugbio_filtering/transformers.py]: maybe I am missing something... not sure why specific encoding is needed for each feature. 
for example,  for featurmap I am using a generalized LabelEncoder to handle this. here is an [example](https://github.com/Ultimagen/ugbio-utils/blob/cd18cd6accad17bb1c938320071fe24e5ed24735/src/featuremap/ugbio_featuremap/somatic_featuremap_inference_utils.py#L30)

File: src/cnv/tests/unit/test_convert_combined_cnv_results_to_output_formats.py
Comment (TammushB) [inline src/cnv/tests/unit/test_convert_combined_cnv_results_to_output_formats.py]: please rename to relevant test name

File: src/core/ugbio_core/vcf_utils.py [hunk 8]
```diff
@@ -482,6 +488,31 @@ def update_vcf_contigs_from_fai(
         # Index the output VCF
         self.index_vcf(output_vcf)
 
+    def remove_filters(
+        self,
+        input_vcf: str,
+        output_vcf: str,
+        filters_to_remove: list[str] | None = None,
+    ) -> None:
+        """Remove specific filters or all filters from a VCF file using bcftools annotate.
```
Comment (TammushB) [inline src/core/ugbio_core/vcf_utils.py]:
```suggestion
        """Remove specific filters or all filters from a VCF records using bcftools annotate.
```

**Review by TammushB — APPROVED** [review]
Ok, looks good I think. 
left some comments.

File: src/cnv/ugbio_cnv/cnv_consts.py
[the remark this answers is not in the document]
Comment (ilyasoifer) [reply src/cnv/ugbio_cnv/cnv_consts.py]: OK

File: src/cnv/ugbio_cnv/cnv_consts.py (hunk 2 above)
Comment (ilyasoifer) [reply src/cnv/ugbio_cnv/cnv_consts.py]: no, in cn.mops it is called CNV_calls_source, just trying not to change too much

File: src/cnv/ugbio_cnv/combine_cnmops_cnvpytor_cnv_calls.py [hunk 9]
```diff
@@ -473,6 +492,8 @@ def filter_dup_cnmmops_cnv_calls(
                     svlen = abs(record.info.get("SVLEN", [0])[0])
                     if svlen < int(filtered_length):
                         if "FILTER" in record.filter.keys():
+                            if "PASS" in record.filter.keys():
+                                record.filter.clear()
                             record.filter.add("CNMOPS_SHORT_DUPLICATION")
                         else:
                             record.filter.add("CNMOPS_SHORT_DUPLICATION")
```
[the remark this answers is not in the document]
Comment (ilyasoifer) [reply src/cnv/ugbio_cnv/combine_cnmops_cnvpytor_cnv_calls.py]: Fixed

File: src/filtering/ugbio_filtering/transformers.py
Comment (ilyasoifer) [reply src/filtering/ugbio_filtering/transformers.py]: I think I prefer it this way

File: src/cnv/ugbio_cnv/cnmops_utils.py (hunk 6 above)
Comment (ilyasoifer) [reply src/cnv/ugbio_cnv/cnmops_utils.py]: I think it is cn.mops specific, in other places we do not create vcf from bed

File: src/cnv/ugbio_cnv/cnmops_utils.py (hunk 7 above)
Comment (ilyasoifer) [reply src/cnv/ugbio_cnv/cnmops_utils.py]: fixed

File: src/cnv/ugbio_cnv/combine_cnv_vcf_utils.py [hunk 10]
```diff
@@ -249,7 +260,15 @@ def merge_cnvs_in_vcf(input_vcf: str, output_vcf: str, distance: int = 1000) ->
...
         Maximum distance between CNV variants to consider them for merging, by default 1000.
-
+    ignore_sv_type : bool, optional
+        Whether to ignore SVTYPE when collapsing variants, by default False.
+    ignore_filter: bool, optional
+        Whether to ignore FILTER status when collapsing variants, by default True.
+    do_not_merge_collapsed_filtered: bool, optional
+        Whether to avoid merging filtered out variants were removed by collapsing, by default False.
+
+    pick_best : bool, optional
```
[the remark this answers is not in the document]
Comment (ilyasoifer) [reply src/cnv/ugbio_cnv/combine_cnv_vcf_utils.py]: added doc
